examples/dynamoathenamigration/migration.go

```diff
@@ -486,18 +489,251 @@ func (t *task) downloadFromS3(ctx context.Context, key string) (*os.File, error)
 	}
 	path := path.Join(dir, originalName)
 
-	f, err := os.Create(path)
+	originalFile, err := os.Create(path)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	if _, err := t.s3Downloader.Download(ctx, f, &s3.GetObjectInput{
+	if _, err := t.s3Downloader.Download(ctx, originalFile, &s3.GetObjectInput{
 		Bucket: aws.String(t.Bucket),
 		Key:    aws.String(key),
 	}); err != nil {
-		f.Close()
+		originalFile.Close()
+		return nil, trace.Wrap(err)
+	}
+
+	defer originalFile.Close()
+
+	// maxMemoryUsedForSortingExportInBytes defines how big chunks of audit events
+	// will be loaded into memory, sorted and stored in temporary files.
+	maxMemoryUsedForSortingExportInBytes := 1024 * 1024 * t.MaxMemoryUsedForSortingExportInMB
+	sortedFile, err := createSortedExport(originalFile, dir, maxMemoryUsedForSortingExportInBytes)
+	if err != nil {
+		return nil, trace.NewAggregate(err, sortedFile.Close())
+	}
+	return sortedFile, nil
+}
+
+type eventWithTime struct {
+	rawLine   json.RawMessage
+	eventDate string
+}
+
+// createSortedExport reads a large json.gz export file, which cannot fit entirely
+// into memory and splits it into multiple smaller sorted (in memory) files.
+// Afterwards, it merges these smaller files into a final sorted output.
+// The function returns json file which is sorted by createdAtDate ascending.
+func createSortedExport(inFile *os.File, dir string, maxSize int) (f *os.File, err error) {
+	tmpSortedFilesDir := path.Join(dir, fmt.Sprintf("sort-%s", uuid.NewString()))
+	if err := os.Mkdir(tmpSortedFilesDir, 0o755); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer func() {
+		err = trace.NewAggregate(err, os.RemoveAll(tmpSortedFilesDir))
+	}()
+
+	gzipReader, err := gzip.NewReader(inFile)
+	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-	return f, nil
+	defer gzipReader.Close()
+
+	dec := json.NewDecoder(gzipReader)
+	events := make([]eventWithTime, 0, 10000)
+	var tmpFiles []string
+	var size int
+	// dec.More read export line by line because export is new line json.
+	for dec.More() {
+		// First decode whole line to rawMessage.
+		var singleLine json.RawMessage
+		err = dec.Decode(&singleLine)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		// Parse just createAtDate.
+		var parsedLine dynamoEventPart
+		err = json.Unmarshal(singleLine, &parsedLine)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		events = append(events, eventWithTime{
+			rawLine:   singleLine,
+			eventDate: parsedLine.Item.CreatedAtDate.Value,
+		})
+		size += len(singleLine)
```

Comment:
Should we include `eventDate` in size calculation? It takes some bytes from the max size.

Reply:
https://github.com/gravitational/teleport/pull/29627/commits/8a14381c6058993ec0ec7e2d2a5841c6d96c36e5